sources/bloodhound/src/main.rs

```diff
@@ -0,0 +1,210 @@
+    /// Path to the directory containing checker binaries [default: /usr/libexec/bloodhound-checks]
+    #[arg(short, long = "checks")]
+    check_dir: Option<PathBuf>,
+    /// Format of the output
+    #[clap(short, long, value_enum, default_value_t=Format::Text)]
+    format: Format,
+    /// The CIS benchmark compliance level to check
+    #[arg(short, long, default_value_t = 1)]
+    level: u8,
+    /// Write output to a file at given path [default: stdout]
+    #[arg(short, long)]
+    output: Option<PathBuf>,
+}
+
+/// Discover all executable checker files for the given directory.
+/// By looking at the provided path, this does some basic checks and filtering
+/// to find the executable files that appear to be bloodhound checkers.
+/// It will parse the checker metadata to provide a mapping between a filesystem
+/// path and a specific check, including only those checkers that are within the
+/// request compliance level.
+fn find_checkers(check_dir: &PathBuf, level: u8) -> HashMap<String, CheckerMetadata> {
+    let entries: Vec<DirEntry> = fs::read_dir(check_dir)
+        .unwrap()
+        .filter_map(|file| file.ok())
+        .collect();
+    let mut result = HashMap::with_capacity(entries.len());
+
+    for entry in entries {
+        if let Ok(file_metadata) = fs::metadata(entry.path()) {
+            // Filter out any subdirectories
+            if file_metadata.is_dir() {
+                continue;
+            }
+
+            // Skip any files that are not executable
+            if file_metadata.permissions().mode() & 0o111 == 0 {
+                continue;
+            }
+
+            // It's an executable file, make sure it implements our expected checker interface
+            let metadata;
+            if let Ok(output) = Command::new(entry.path()).arg("metadata").output() {
+                metadata = String::from_utf8_lossy(&output.stdout).to_string();
+            } else {
+                eprintln!(
+                    "{:?} does not appear to be a checker executable",
+                    entry.path()
+                );
+                continue;
+            }
+
+            if let Ok(checker_data) = serde_json::from_str::<CheckerMetadata>(&metadata) {
+                if checker_data.level <= level {
+                    result.insert(
+                        // Assuming here that we will never have non-ASCII characters in our usage
+                        entry.path().as_os_str().to_string_lossy().to_string(),
+                        checker_data,
+                    );
+                }
+            } else {
+                eprintln!("Unable to parse checker metadata from {:?}", entry.path());
+            }
+        }
+    }
+
+    result
+}
+
+/// Processes the output results from calling an individual checker. Results of
+/// parsing the checker output are added to the overall execution report.
+fn process_checker_results(output: Output, data: CheckerMetadata, report: &mut ReportResults) {
```

Comment:
nit: Why not handling this in `add_result`? You will save yourself the extra `mut` reference.